### ISSUE TITLE
Confusing `levels` fallback

### ISSUE THREAD

**Issue by aplavin** [JuliaData/DataAPI.jl#54]
Looking at the concrete implementation of `levels` in `CategoricalArrays`, I see that this function conveniently extracts possible levels from both arrays and individual values:
```julia
julia> a = CategoricalArray(["abc", "def", "abc"])
3-element CategoricalArray{String,1,UInt32}:
 "abc"
 "def"
 "abc"

julia> x = a[1]
CategoricalValue{String, UInt32} "abc"

julia> levels(a)
2-element Vector{String}:
 "abc"
 "def"

julia> levels(x)
2-element Vector{String}:
 "abc"
 "def"
```

However, the fallback implemented in `DataAPI` itself is only correct for collections, not individual values:
```julia
# like unique(x), makes sense?
julia> levels(["abc", "def", "abc"])
2-element Vector{String}:
 "abc"
 "def"

# makes no sense:
julia> levels("abc")
3-element Vector{Char}:
 'a': ASCII/Unicode U+0061 (category Ll: Letter, lowercase)
 'b': ASCII/Unicode U+0062 (category Ll: Letter, lowercase)
 'c': ASCII/Unicode U+0063 (category Ll: Letter, lowercase)

# especially confusing given that:
julia> x == "abc"
true
```

Maybe, the fallback shouldn't exist at all, and something like `haslevels(x)::Bool` added instead?

**Comment by bkamins** [comment]
> is only correct for collections

The problem is different. The function is correct, but it is defined for all collections, e.g.:
```
julia> levels("abc")
3-element Vector{Char}:
 'a': ASCII/Unicode U+0061 (category Ll: Letter, lowercase)
 'b': ASCII/Unicode U+0062 (category Ll: Letter, lowercase)
 'c': ASCII/Unicode U+0063 (category Ll: Letter, lowercase)
```
The problem is, that many collections, like `"abc"` or `1` are perceived by users as individual values not as collections.

The same problem is with `unique`.

What we maybe could do is changing that `levels` is defined for arrays by default (the problem is that it would be breaking).

**Comment by bkamins** [comment]
Let me summarize the discussion, so that I am clear what problem we want to resolve (I think you raised it in Slack but I do not remember the details, and soon it will be lost on Slack). Current state is the following:
* for collections `levels` recovers levels of a collection;
* for `CategoricalValue` as a special case `levels` recovers `levels` from a source array of this categorical value.

In what cases the current behavior is problematic (I mean in practice - as I understand your conceptual concerns).

**Comment by bkamins** [comment]
> one needs to know in advance that the input is a `CategoricalValue`, otherwise `levels(val)` don't make much sense.

I understand the current design the opposite way (I do not want to say that it should not be changed, but I want to express my current understanding of the design):
* `levels` normally should be only called on a collection; you should not call it on a value that is not a collection
* as a special case, if you know that you have a `CategoricalValue` you can call it to get `levels` of a parent of this value

So, you should not call `levels` on any value that is not a collection, unless you know beforehand that it is a `CategoricalValue`.

Therefore, what you essentially ask for is how to check that a non-collection is `CategoricalValue` without having CategoricalArrays.jl as a dependency. This could be doable (e.g. we could add some DataAPI.jl method to check for it). However, a normal way to do it would be to add dependency on CategoricalArrays.jl and just check if some value is a `CategoricalValue`.

An alternative would be to add something like `parentlevels` function, so that `parentlevels` would be defined for `CategoricalValue` as the same as `levels`, and would be `nothing` by default as you want. @nalimilan - what do you think?

**Comment by bkamins** [comment]
Agreed. That is why maybe adding `parentlevels` to be explicit makes sense. Let us wait for @nalimilan to respond what he things, as he designed these features.

**Comment by nalimilan** [comment]
Yes, I think the best solution would be to add an API that allows checking whether a value or array is categorical, whose only implementations (currently) would be `CategoricalArray` and `CategoricalValue`.

It's true that the fact that the `levels` fallback works on any iterable value and can return weird results for strings and numbers is a bit confusing, but introducing `parentlevels` would be overkill IMO.

**Comment by bkamins** [comment]
And for the mean time having CategoricalArrays.jl as a dependency is a temporary solution.

**Comment by nalimilan** [comment]
See https://github.com/JuliaData/CategoricalArrays.jl/issues/200 about adding `iscategorical`.
